src/core/radio/ble_secure.cpp

```diff
@@ -0,0 +1,475 @@
+
+    SuccessOrAssert(mTls.SetPsk(reinterpret_cast<const uint8_t *>(aPskd.GetAsCString()), aPskd.GetLength()));
+}
+
+Error BleSecure::SendMessage(ot::Message *aMessage)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+    VerifyOrExit(aMessage != nullptr, error = kErrorInvalidArgs);
+    SuccessOrExit(error = mSentMessage->AppendBytesFromMessage(*aMessage, 0, aMessage->GetLength()));
+
+    if (mSentMessage->GetLength() > kFlushThreshold)
+    {
+        error = Flush();
+    }
+
+    if (error == kErrorNone)
+        FreeMessage(aMessage);
+
+exit:
+    return error;
+}
+
+Error BleSecure::Send(uint8_t *aBuf, uint16_t aLength)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+    SuccessOrExit(error = mSentMessage->AppendBytes(aBuf, aLength));
+
+    if (mSentMessage->GetLength() > kFlushThreshold)
+    {
+        error = Flush();
+    }
+
+exit:
+    return error;
+}
+
+Error BleSecure::SendApplicationTlv(uint8_t *aBuf, uint16_t aLength)
+{
+    if (aLength > 254)
+    {
+        ot::ExtendedTlv tlv;
+
+        tlv.SetType(ot::MeshCoP::TcatAgent::TlvType::kSendApplicationData);
+        tlv.SetLength(aLength);
+        Send(reinterpret_cast<uint8_t *>(&tlv), sizeof(tlv));
+    }
+    else
+    {
+        ot::Tlv tlv;
+
+        tlv.SetType(ot::MeshCoP::TcatAgent::TlvType::kSendApplicationData);
+        tlv.SetLength(aLength);
+        Send(reinterpret_cast<uint8_t *>(&tlv), sizeof(tlv));
+    }
+
+    return Send(aBuf, aLength);
+}
+
+Error BleSecure::Flush()
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+    VerifyOrExit(mSentMessage->GetLength() != 0, error = kErrorNone);
+
+    mTransmitQueue.Enqueue(*mSentMessage);
+    mTransmitTask.Post();
+
+    mSentMessage = Get<MessagePool>().Allocate(Message::kTypeBle);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+
+exit:
+    return error;
+}
+
+Error BleSecure::HandleBleReceive(uint8_t *aBuf, uint16_t aLength)
+{
+    ot::Message     *message = nullptr;
+    Ip6::MessageInfo theMessageInfo;
+
+    VerifyOrExit((message = Get<MessagePool>().Allocate(Message::kTypeBle, 0)) != nullptr);
+    SuccessOrExit(message->AppendBytes(aBuf, aLength));
+
+    // Cannot call Receive(..) directly because Setup(..) and mState are private
+    mTls.HandleUdpReceive(*message, theMessageInfo);
+
+    FreeMessage(message);
+
+    return kErrorNone;
+
+exit:
+    FreeMessage(message);
+    return kErrorNoBufs;
+}
+
+Error BleSecure::HandleBleConnected(uint16_t aConnectionId)
+{
+    OT_UNUSED_VARIABLE(aConnectionId);
+
+    mBleConnectionOpen = true;
+
+    otPlatBleGattMtuGet(&GetInstance(), &mMtuSize);
```

Comment:
Do we know the MTU size at this exact moment (i.e. right after connection is made)? Normally, central (client) needs to send the Excchange MTU request to inform about it's MTU and asks about MTU of peripheral (if it's non 23).

Reply:
I added a new API "otPlatBleGattOnMtuUpdate(...)" for the radio driver to inform OpenThread about MTU updates.